### DISCUSSION TITLE
Generic CIP to Turck IO Link Device

### DISCUSSION THREAD

**Discussion by sdmolt** [ottowayi/pycomm3#177] [Q&A] [answered]
I'm trying to figure out how to read IO link parameter data using `generic_message`. 

The IO-Link master is a [TBEN-S2-4IOL](https://www.turck.de/attachment/D301369.pdf) and my device is the [TBIL-M1-16DXP](https://www.turck.de/attachment/100014026.pdf). 

The parameter I'm interested in is 64 "Parameter ID" (page 13 in the TBIL manual). On page 83 in the TBEN manual, it has an example on how to get similar data. I've tried quite a few different combinations, but I always get "Service not supported". This is what I think should be correct:

```
from pycomm3 import UDINT
from pycomm3 import CIPDriver

ip = '192.168.1.110'
with CIPDriver(ip) as IOMaster:
    param = IOMaster.generic_message(
        service=b'\x4B',
        class_code=b'\x67',
        instance=b'\x01',
        attribute=b'\x01',
        request_data=b'\x40',
        data_type=UDINT,
        name='ReadParam',
        connected=False,
        unconnected_send=True,
        route_path=True
    )
    print(param)
```

Thanks!

**Comment by ottowayi** [comment, marked as answer]
I have no experience with IO-Link nor these devices, but I'll try my best to give you some tips.  I would start by removing the `unconnected_send` and `route_path` arguments, since you're connecting to this device directly there is nothing to route so those shouldn't be needed.  Setting `connected` to False is also probably what you want.  I would also make sure you're on the latest version, there was a bug that was adding to the CIP route in the CIPDriver when it should only happen in the LogixDriver.   The `service`, `class`, `instance`, and `attribute` args all look correct, but the `request_data` you only have 1-byte and the example shows 3: Index (LSB), Index (MSB), Sub Index.  I think your Index (`0x40`) needs to be an UINT or 2-bytes and you're missing the sub-index.  Can you try `request_data=UINT.encode(0x40) + b'\x00`?

**Reply by sdmolt** [reply]
Thanks for the response! The request data did it - I was able to get it working correctly. 

```
from pycomm3 import UDINT
from pycomm3 import CIPDriver
ip = '192.168.1.110'

with CIPDriver(ip) as IOMaster:
	read_isdu = IOMaster.generic_message(
		service=b'\x4B',                
		class_code=b'\x67',             
		instance=b'\x01',               
		attribute=b'\x01',              
		request_data=b'\x40\x00\x00',   
		data_type=UDINT,                
		connected=False                 
	)

	parameter0x40 = UDINT.decode(read_isdu[1].to_bytes(4, 'big'))

	print(parameter0x40)
```

**Reply by dealeson** [reply]
Hi!

I have a simmilar issue.
I am communicating via EthernetIP with another IO-Link Master device, using generic_message().

There's no problem in reading index/subindex, but the problem arrives when dealing with write data.

No matter what I define as "request_data" ,  the library adds two 0x00 bytes at the end of the message, so by default I get an error , and i can't write into index/subindex.

EXAMPLE:
`
with CIPDriver(ip) as IOLink:
    write_isdu = IOLink.generic_message(               
        class_code=b'\x80',            
        instance=b'\x01',                
        attribute=b'\x01', 
        service=b'\x4C',              
        request_data=b'\xe0\x2e\x01\x02',                 
        connected=False , 
        unconnected_send=False  ,    
    )
    print(write_isdu)
`

TRANSMITTED (Only final CIP Data )
`
Data: e02e01020000 <-- it should be e02e0102
`
CONSOLE:
`generic, b'\x803', None, Request service error`

As far as I'm a python experienced programmer (only user), I am not able to find the solution, i was searching into the classes definitions but i couldn't find anything to fix it.

could anyone check this?

The issue, by reading data, it's not a problem because it seems that those extra "00's" are irrelevant in EthIP/CIP communications....
As a try, i've inserted a lots of \x00 in the end of the request_data (when reading) and has no effect in received_data

Obviously I've test this equipment with another software and it did the trick:
Software used: EthernetIP tools by Molex , it handles with no problem reading and writing data (but I need use python to develop a solution)
